### DISCUSSION TITLE
Value vs reference types

### DISCUSSION THREAD

**Discussion by mansellan** [twinbasic/twinbasic#931] [General]
In garbage-collected environments, for performance-critical code, it's important to try to do as much as possible using stack-allocated types (value types, including arrays of them if specifically marked) rather than heap-allocated types (reference types). However, this is mostly due to the cost of garbage collection - allocation on the heap is cheap, but their deallocation (cleanup) is not.

VBx and twinBASIC are reference-counted, so I would expect that deallocation on the heap is much cheaper (almost free?). However, instinctively I expect that even reference-counted heap (de)allocations could not be as cheap as stack-allocated. For example, does VBx / twinBASIC ever need to defragment the heap? If so, when does it happen and how expensive is it?

The reason I ask is that if heap (de)allocation is sufficiently performant, there's probably no need to consider making value types more expressive (think UDT methods).

Obviously none of this matters for v1, or indeed for 99.99% of users. But it may be of concern for library / package authors.

**Comment by mburns08109** [comment]
IIRC, the COM world offered no analog to garbage collection on the heap, and free memory fragmentation was one of the direct results.
Sometimes you just could not get large contiguous blocks of free memory (at least not without virtual page creation/swapping as a necessary cost). This was one of the bigger issues with COM, in the end which led them towards the whole .NET thing in the first place (well, that and of course the registry issues).
Sometimes you just had to wait for enough free ram to open up as other programs ended/unloaded.
